### DISCUSSION TITLE
Want to build a Prisma extension for `@default(<fn>)`, could you hint where I should look?

### DISCUSSION THREAD

**Discussion by grimen** [prisma/prisma#6930] [Q&A] [answered]
I have a very clear goal in mind of trying to make an extension to Prisma to support custom `@default(<customFunction>())` functions defined in environment, e.g. P/SQL. 

In my case I have a custom ULID function written in P/SQL that generated PostgreSQL UUID-compatible sequential UUIDs, which I would want to define in a Prisma schema as `@default(ulid())`.

I searched the code base with GitHub Search for hints on how the existing functionality is implemented, but I didn't find any matching source.

Thus, my question is: Could you share some pointers on where the native implementation of the `fn` in `@default(<fn>())` is located?

Thanks in advance

**Comment by janpio** [comment, marked as answer]
What exactly are you trying to achieve? The existing Prisma level functions like `uuid()` or `cuid()` are pretty deeply embedded - and actually run on the Query Engine side (and not the database). If you have something that you could store as a Postgres function, you can already use those in as the default by defining them as a string default. Is that what you are looking for?

**Reply by grimen** [reply]
For database index performance reasons when using UUIDs. The background of my prioritized challenge can be explained by one or a few of these links:

- https://www.2ndquadrant.com/en/blog/sequential-uuid-generators
- https://stackoverflow.com/questions/63057079/ulid-native-support-in-postgresql
- https://github.com/tvondra/sequential-uuids
- https://github.com/iCyberon/pg_ulid
- https://github.com/ulid/spec

But generally speaking, technically since in raw SQL the default value could be *anything* I think supporting custom functions defined in the database runtime environment would be very useful for multiple other cases.

I do know that Prisma has "hooks", but I personally would prefer default values to be interpreted in the actual database environment so schemas are more consistent.

**Reply by grimen** [reply]
Oh regarding your last sentence, do you have examples of this? I could not find / understand that aspect from the documentation.

**Reply by janpio** [reply]
If the function is defined in the database, then this already works. 

There is a difference between `@default(uuid())` which is a Prisma level thing (and hard to adapt - the existing functions are listed here: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#attribute-functions), and `@default("whatevershouldexecutedinthesql")` which really lets the database do whatever it needs to do with that String when migrating your table. 

I think you are actually looking for the latter, so you can define the function in your database and just let Prisma create a column with a default value that executes that database level function. (Our documentation around this is pretty crappy right now, I am trying to find something that help you along here.)

A good way to understand this would be to just create your database with a table manually and set a Postgres level function as the default. Then run `npx prisma db pull` and see how this is represented in the schema.

**Reply by grimen** [reply]
Oh awesome, this helps a lot and from what I can understand resolves my challenge. I have a `next_uuid` function in my case defined in database environment, will try it out later. Thanks! =D

**Reply by janpio** [reply]
Please keep me updated how it goes. We can then also use this conversation here to write the correct documentation, and also hopefully help resolve any problems you might hit.

**Reply by janpio** [reply]
Ok, I found one example in the schema documentation, which shows I got it slightly wrong above: 
https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated
```
model User {
  id   String  @id @db.Uuid @default(dbgenerated("gen_random_uuid()"))
  test String?
}
```

So you will need to use `dbgenerated` around the function you want to call.

**Reply by akshayvadher** [reply]
What if my DB does not have a function to generate ULID (or KSUID or NanoId), in that case, I would want to rely on a javascript function (custom made in somecasese) to generate those ids. 

Is there a way in prisma to call a custom javascript function instead of predefined? 

Or please provide pointer to do it, since `uuid`, `cuid`, or `now` are prisma defined functions, there is potentially a way to inject custom functions as well.

**Reply by GermanJablo** [reply]
> What if my DB does not have a function to generate ULID (or KSUID or NanoId), in that case, I would want to rely on a javascript function (custom made in somecasese) to generate those ids.
> 
> Is there a way in prisma to call a custom javascript function instead of predefined?
> 
> Or please provide pointer to do it, since uuid, cuid, or now are prisma defined functions, there is potentially a way to inject custom functions as well.

I have exactly the same question. Any ideas?

**Reply by ruheni** [reply]
Hey 👋

You could achieve this with a [Prisma Client extension](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query) by intercepting queries and creating the id

Given the following schema:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model User {
  id    String  @id @default("")
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id        String  @id @default("")
  title     String
  content   String?
  published Boolean @default(false)
  author    User?   @relation(fields: [authorId], references: [id])
  authorId  String?
}

```

You can use Prisma Client extensions to intercept queries and define your own id before executing a query. In this case, I'm using `nanoid` and applying it to all my models.

```js
import { PrismaClient } from '@prisma/client'
import { nanoid } from 'nanoid'

const prisma = new PrismaClient()
  .$extends({
    query: {
      $allModels: {
        async create({ args, query }) {
          args.data = {
            ...args.data,
            id: nanoid(),
          }

          return query(args)
        }
      }
    }

  })

const post = await prisma.post.create({
  data: {
    title: 'random title',
  }
})

// Result
{
  id: 'hMEny4U_XNQuMLDYTjl97',
  title: 'random title',
  content: null,
  published: false,
  authorId: null
}
```

**Reply by AndreyMarchuk** [reply]
don't forget `createMany()`, `createManyAndReturn()`, `upsert()`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by HokaigoTwin** [comment]
Hey everyone,

I was actually just reading through the Prisma docs recently, noticed we couldn't pass custom JS functions into `@default()`, and thought it would be a cool feature to build. 

Looking at the manual `$extends` examples discussed above, it definitely gets messy once you try to handle `createMany`, `upsert`, or deeply nested relational writes without hardcoding model names everywhere. 

So, I put together a zero-dependency package that parses the DMMF and handles this automatically via schema comments. Just wanted to drop it here for anyone looking for a cleaner setup!

Instead of writing massive middlewares, you just do this:

**1. Schema:**
```prisma
model User {
  /// @defaultJs(generateCustomId)
  id String @id @default("")
  email String @unique
}
```

**2. Extension:**
```typescript
import { PrismaClient } from '@prisma/client';
import { withJsDefaults } from 'prisma-js-defaults';
// Import the config auto-generated by the package
import { jsDefaultsConfig } from './prisma/generated/js-defaults';

const prisma = new PrismaClient().$extends(
    withJsDefaults(jsDefaultsConfig, {
        generateCustomId: () => `user_${nanoid()}`
    })
);
```
It intercepts all operations (including nested creations) and applies your JS functions out of the box. 

You can check it out here:
https://github.com/HokaigoTwin/prisma-js-defaults
